File: spec/Woot/WStringSpec.hs
module Data.Woot.WStringSpec where


import MockData (mockWString)

import Data.Woot.WChar
import Data.Woot.WString
import Test.Hspec


spec :: Spec
spec = do
    describe "show" $
        it "should convert a woot string to a generic string" $
            show mockWString `shouldBe` "bar"

    describe "insert" $
        it "should insert a character before the specified index" $
            let newChar = WChar (WCharId 0 4) True  'P' (WCharId 0 1) (WCharId 0 2) in
            show (insert newChar 4 mockWString) `shouldBe` "baPr"

    describe "subsection" $ do
        it "should return a subsection specified by the provided ids" $ do
            show (subsection (WCharId 5 5) (WCharId 7 7) mockWString) `shouldBe` ""
            show (subsection (WCharId 0 1) (WCharId (-1) 1) mockWString) `shouldBe` "ar"

        it "should return nothing when passed valid ids that produce an invalid splice" $
            show (subsection (WCharId 0 (-1)) (WCharId 0 1) mockWString) `shouldBe` ""

File: src/woot/wstring.spec.js
import {
  show,
  insert,
  subsection,
} from './wstring';

import { mockWString } from '../../spec/mock-data';

describe('WString', () => {
  describe('show', () => {
    it('should convert a woot string to a generic string', () => {
      expect(show(mockWString)).toBe('bar');
    });
  });

  describe('insert', () => {
    it('should insert a character before the specified index', () => {
      const newChar = {
        id: { clientId: 0, clock: 4 },
        isVisible: true,
        alpha: 'P',
        prevId: { clientId: 0, clock: 1 },
        nextId: { clientId: 0, clock: 2 },
      };

      const newString = insert(4, newChar, mockWString);

      expect(show(newString)).toBe('baPr');
    });
  });

  describe('subsection', () => {
    it('should return an empty WString when passed ids that do not make a subsection', () => {
      const section = subsection(
        { clientId: -1, clock: 1 },
        { clientId: 0, clock: 1 },
        mockWString
      );

      expect(section).toEqual([]);
    });

    it('should return a subsection when passed valid ids', () => {
      const section = subsection(
        { clientId: 0, clock: 1 },
        { clientId: -1, clock: 1 },
        mockWString
      );

      expect(show(section)).toEqual('ar');
    });
  });
});
